### PR TITLE
PHPC-851: Separate argument zvals before modification

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-851

### CODE REVIEW

File: php_phongo.c [hunk 1]
```diff
@@ -837,7 +837,9 @@ void php_phongo_read_preference_prep_tagsets(zval *tagSets TSRMLS_DC) /* {{{ */
 		zval *tagSet;
 
 		ZEND_HASH_FOREACH_VAL(ht_data, tagSet) {
+			ZVAL_DEREF(tagSet);
```
Comment (derickr) [inline php_phongo.c]: Why do you do the deref as well?

**Comment by bashofmann** [comment]
Looks good, with this patch I can't reproduce the segmentation fault on setting read preferences anymore

File: php_phongo.c (hunk 1 above)
Comment (jmikola) [inline php_phongo.c]: Basing this on iteration examples I found in core (e.g. [`php_zval_filter_recursive()`](https://github.com/php/php-src/blob/php-7.0.14/ext/filter/filter.c#L508)).

We currently handle dereferencing in `phongo_bson_append()`, which is the only other time we iterate apart from preparing tag sets. BSON encoding also resolves indirect references, which I could do with the `ZEND_HASH_FOREACH_VAL_IND()` macro, but I wasn't sure if that was necessary (even after reading up on indirect zvals in [Nikita's blog](https://nikic.github.io/2015/06/19/Internal-value-representation-in-PHP-7-part-2.html)).

Comment (jmikola) [inline php_phongo.c]: I opened [PHPC-870](https://jira.mongodb.org/browse/PHPC-870) to track future iteration changes.

**Pull request merged**